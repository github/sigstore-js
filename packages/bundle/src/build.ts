/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { HashAlgorithm } from '@sigstore/protobuf-specs';
import { BUNDLE_V02_MEDIA_TYPE, BUNDLE_V03_MEDIA_TYPE } from './bundle';

import type { Envelope, Signature } from '@sigstore/protobuf-specs';
import type {
  Bundle,
  BundleWithDsseEnvelope,
  BundleWithMessageSignature,
} from './bundle';

// Helper functions for assembling the parts of a Sigstore bundle

type VerificationMaterialOptions = {
  certificate?: Buffer;
  keyHint?: string;

  // When set to true, the bundle verification material will use the
  // certifciate field instead of the x509CertificateChain field.
  // When undefied/false, a v0.2 bundle will be created.
  singleCertificate?: boolean;
};

type MessageSignatureBundleOptions = {
  digest: Buffer;
  signature: Buffer;
} & VerificationMaterialOptions;

type DSSEBundleOptions = {
  artifact: Buffer;
  artifactType: string;
  signature: Buffer;
} & VerificationMaterialOptions;

// Message signature bundle - $case: 'messageSignature'
export function toMessageSignatureBundle(
  options: MessageSignatureBundleOptions
): BundleWithMessageSignature {
  return {
    mediaType: options.singleCertificate
      ? BUNDLE_V03_MEDIA_TYPE
      : BUNDLE_V02_MEDIA_TYPE,
    content: {
      $case: 'messageSignature',
      messageSignature: {
        messageDigest: {
          algorithm: HashAlgorithm.SHA2_256,
          digest: options.digest,
        },
        signature: options.signature,
      },
    },
    verificationMaterial: toVerificationMaterial(options),
  };
}

// DSSE envelope bundle - $case: 'dsseEnvelope'

export function toDSSEBundle(
  options: DSSEBundleOptions
): BundleWithDsseEnvelope {
  return {
    mediaType: options.singleCertificate
      ? BUNDLE_V03_MEDIA_TYPE
      : BUNDLE_V02_MEDIA_TYPE,
    content: {
      $case: 'dsseEnvelope',
      dsseEnvelope: toEnvelope(options),
    },
    verificationMaterial: toVerificationMaterial(options),
  };
}

function toEnvelope(options: DSSEBundleOptions): Envelope {
  return {
    payloadType: options.artifactType,
    payload: options.artifact,
    signatures: [toSignature(options)],
  };
}

function toSignature(options: DSSEBundleOptions): Signature {
  return {
    keyid: options.keyHint || '',
    sig: options.signature,
  };
}

// Verification material

function toVerificationMaterial(
  options: VerificationMaterialOptions
): Bundle['verificationMaterial'] {
  return {
    content: toKeyContent(options),
    tlogEntries: [],
    timestampVerificationData: { rfc3161Timestamps: [] },
  };
}

function toKeyContent(
  options: VerificationMaterialOptions
): Bundle['verificationMaterial']['content'] {
  if (options.certificate) {
    if (options.singleCertificate) {
      return {
        $case: 'certificate',
        certificate: { rawBytes: options.certificate },
      };
    } else {
      return {
        $case: 'x509CertificateChain',
        x509CertificateChain: {
          certificates: [{ rawBytes: options.certificate }],
        },
      };
    }
  } else {
    return {
      $case: 'publicKey',
      publicKey: {
        hint: options.keyHint || '',
      },
    };
  }
}
