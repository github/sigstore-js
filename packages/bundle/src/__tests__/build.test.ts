/*
Copyright 2023 The Sigstore Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
import { Bundle } from '@sigstore/protobuf-specs';
import { toDSSEBundle, toMessageSignatureBundle } from '../build';

const signature = Buffer.from('signature');
const keyHint = 'hint';
const certificate = Buffer.from('certificate');

describe('toMessageSignatureBundle', () => {
  const digest = Buffer.from('digest');

  describe('when the singleCertificate option is undefined', () => {
    it('returns a valid message signature bundle', () => {
      const b = toMessageSignatureBundle({
        digest,
        signature,
        certificate,
        keyHint,
      });

      expect(b).toBeTruthy();
      expect(Bundle.toJSON(b)).toMatchSnapshot();
    });
  });

  describe('when the singleCertificate option is true', () => {
    it('returns a valid message signature bundle', () => {
      const b = toMessageSignatureBundle({
        digest,
        signature,
        certificate,
        keyHint,
        singleCertificate: true,
      });

      expect(b).toBeTruthy();
      expect(Bundle.toJSON(b)).toMatchSnapshot();
    });
  });
});

describe('toDSSEBundle', () => {
  const artifact = Buffer.from('data');
  const artifactType = 'text/plain';

  describe('when the singleCertificate option is undefined/false', () => {
    describe('when a public key w/ hint is provided', () => {
      it('returns a valid DSSE bundle', () => {
        const b = toDSSEBundle({
          artifact,
          artifactType,
          signature,
          keyHint,
        });

        expect(b).toBeTruthy();
        expect(Bundle.toJSON(b)).toMatchSnapshot();
      });
    });

    describe('when a public key w/o hint is provided', () => {
      it('returns a valid DSSE bundle', () => {
        const b = toDSSEBundle({
          artifact,
          artifactType,
          signature,
          singleCertificate: false,
        });

        expect(b).toBeTruthy();
        expect(Bundle.toJSON(b)).toMatchSnapshot();
      });
    });

    describe('when a certificate chain provided', () => {
      it('returns a valid DSSE bundle', () => {
        const b = toDSSEBundle({
          artifact,
          artifactType,
          signature,
          certificate,
        });

        expect(b).toBeTruthy();
        expect(Bundle.toJSON(b)).toMatchSnapshot();
      });
    });
  });

  describe('when the singleCertificate option is true', () => {
    describe('when a public key w/ hint is provided', () => {
      it('returns a valid DSSE bundle', () => {
        const b = toDSSEBundle({
          artifact,
          artifactType,
          signature,
          keyHint,
          singleCertificate: true,
        });

        expect(b).toBeTruthy();
        expect(Bundle.toJSON(b)).toMatchSnapshot();
      });
    });

    describe('when a public key w/o hint is provided', () => {
      it('returns a valid DSSE bundle', () => {
        const b = toDSSEBundle({
          artifact,
          artifactType,
          signature,
          singleCertificate: true,
        });

        expect(b).toBeTruthy();
        expect(Bundle.toJSON(b)).toMatchSnapshot();
      });
    });

    describe('when a certificate chain provided', () => {
      it('returns a valid DSSE bundle', () => {
        const b = toDSSEBundle({
          artifact,
          artifactType,
          signature,
          certificate,
          singleCertificate: true,
        });

        expect(b).toBeTruthy();
        expect(Bundle.toJSON(b)).toMatchSnapshot();
      });
    });
  });
});
