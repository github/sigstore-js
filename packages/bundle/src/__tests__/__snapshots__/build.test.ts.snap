// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toDSSEBundle when the singleCertificate option is true when a certificate chain provided returns a valid DSSE bundle 1`] = `
{
  "dsseEnvelope": {
    "payload": "ZGF0YQ==",
    "payloadType": "text/plain",
    "signatures": [
      {
        "keyid": "",
        "sig": "c2lnbmF0dXJl",
      },
    ],
  },
  "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
  "verificationMaterial": {
    "certificate": {
      "rawBytes": "Y2VydGlmaWNhdGU=",
    },
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
  },
}
`;

exports[`toDSSEBundle when the singleCertificate option is true when a public key w/ hint is provided returns a valid DSSE bundle 1`] = `
{
  "dsseEnvelope": {
    "payload": "ZGF0YQ==",
    "payloadType": "text/plain",
    "signatures": [
      {
        "keyid": "hint",
        "sig": "c2lnbmF0dXJl",
      },
    ],
  },
  "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
  "verificationMaterial": {
    "publicKey": {
      "hint": "hint",
    },
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
  },
}
`;

exports[`toDSSEBundle when the singleCertificate option is true when a public key w/o hint is provided returns a valid DSSE bundle 1`] = `
{
  "dsseEnvelope": {
    "payload": "ZGF0YQ==",
    "payloadType": "text/plain",
    "signatures": [
      {
        "keyid": "",
        "sig": "c2lnbmF0dXJl",
      },
    ],
  },
  "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
  "verificationMaterial": {
    "publicKey": {
      "hint": "",
    },
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
  },
}
`;

exports[`toDSSEBundle when the singleCertificate option is undefined/false when a certificate chain provided returns a valid DSSE bundle 1`] = `
{
  "dsseEnvelope": {
    "payload": "ZGF0YQ==",
    "payloadType": "text/plain",
    "signatures": [
      {
        "keyid": "",
        "sig": "c2lnbmF0dXJl",
      },
    ],
  },
  "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
  "verificationMaterial": {
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
    "x509CertificateChain": {
      "certificates": [
        {
          "rawBytes": "Y2VydGlmaWNhdGU=",
        },
      ],
    },
  },
}
`;

exports[`toDSSEBundle when the singleCertificate option is undefined/false when a public key w/ hint is provided returns a valid DSSE bundle 1`] = `
{
  "dsseEnvelope": {
    "payload": "ZGF0YQ==",
    "payloadType": "text/plain",
    "signatures": [
      {
        "keyid": "hint",
        "sig": "c2lnbmF0dXJl",
      },
    ],
  },
  "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
  "verificationMaterial": {
    "publicKey": {
      "hint": "hint",
    },
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
  },
}
`;

exports[`toDSSEBundle when the singleCertificate option is undefined/false when a public key w/o hint is provided returns a valid DSSE bundle 1`] = `
{
  "dsseEnvelope": {
    "payload": "ZGF0YQ==",
    "payloadType": "text/plain",
    "signatures": [
      {
        "keyid": "",
        "sig": "c2lnbmF0dXJl",
      },
    ],
  },
  "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
  "verificationMaterial": {
    "publicKey": {
      "hint": "",
    },
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
  },
}
`;

exports[`toMessageSignatureBundle when the singleCertificate option is true returns a valid message signature bundle 1`] = `
{
  "mediaType": "application/vnd.dev.sigstore.bundle.v0.3+json",
  "messageSignature": {
    "messageDigest": {
      "algorithm": "SHA2_256",
      "digest": "ZGlnZXN0",
    },
    "signature": "c2lnbmF0dXJl",
  },
  "verificationMaterial": {
    "certificate": {
      "rawBytes": "Y2VydGlmaWNhdGU=",
    },
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
  },
}
`;

exports[`toMessageSignatureBundle when the singleCertificate option is undefined returns a valid message signature bundle 1`] = `
{
  "mediaType": "application/vnd.dev.sigstore.bundle+json;version=0.2",
  "messageSignature": {
    "messageDigest": {
      "algorithm": "SHA2_256",
      "digest": "ZGlnZXN0",
    },
    "signature": "c2lnbmF0dXJl",
  },
  "verificationMaterial": {
    "timestampVerificationData": {
      "rfc3161Timestamps": [],
    },
    "tlogEntries": [],
    "x509CertificateChain": {
      "certificates": [
        {
          "rawBytes": "Y2VydGlmaWNhdGU=",
        },
      ],
    },
  },
}
`;
